go/database/mpt/codes.go

```diff
@@ -0,0 +1,343 @@
+	"golang.org/x/crypto/sha3"
+)
+
+// codes is a simple data structure to store and manage the codes of accounts.
+// All codes are retained in memory, incrementally backed up to disk during
+// checkpoint and flush operations.
+type codes struct {
+	codes    map[common.Hash][]byte // < all managed codes
+	pending  []common.Hash          // < the hashes of codes not written to disk yet
+	file     string                 // < the file to store the codes
+	fileSize uint64                 // < the current file size
+	mutex    sync.Mutex
+	hasher   hash.Hash
+
+	directory  string                // < a directory for placing checkpoint data
+	checkpoint checkpoint.Checkpoint // < the last checkpoint
+}
+
+var emptyCodeHash = common.GetHash(sha3.NewLegacyKeccak256(), []byte{})
+
+const (
+	fileNameCodes                    = "codes.dat"
+	fileNameCodesCheckpointDirectory = "codes"
+	fileNameCodesCommittedCheckpoint = "committed.json"
+	fileNameCodesPrepareCheckpoint   = "prepare.json"
+)
+
+func openCodes(stateDirectory string) (*codes, error) {
+	file, directory := getCodePaths(stateDirectory)
+	if err := os.MkdirAll(directory, 0700); err != nil {
+		return nil, err
+	}
+
+	// Create the code file if it does not exist.
+	if _, err := os.Stat(file); os.IsNotExist(err) {
+		if err := os.WriteFile(file, []byte{}, 0600); err != nil {
+			return nil, err
+		}
+	}
+
+	data, size, err := readCodesAndSize(file)
+	if err != nil {
+		return nil, err
+	}
+
+	committed := filepath.Join(directory, fileNameCodesCommittedCheckpoint)
+	meta, err := readCodeCheckpointMetaData(committed)
+	if err != nil {
+		return nil, err
+	}
+
+	return &codes{
+		codes:      data,
+		file:       file,
+		fileSize:   size,
+		directory:  directory,
+		hasher:     sha3.NewLegacyKeccak256(),
+		checkpoint: meta.Checkpoint,
+	}, nil
+}
+
+func (c *codes) add(code []byte) common.Hash {
+	hash := common.GetHash(c.hasher, code)
+	c.mutex.Lock()
+	if _, found := c.codes[hash]; !found {
+		c.codes[hash] = code
+		c.pending = append(c.pending, hash)
+	}
+	c.mutex.Unlock()
+	return hash
+}
+
+func (c *codes) getCodeForHash(hash common.Hash) []byte {
+	c.mutex.Lock()
+	res := c.codes[hash]
+	c.mutex.Unlock()
+	return res
+}
+
+func (c *codes) getCodes() map[common.Hash][]byte {
+	c.mutex.Lock()
+	res := maps.Clone(c.codes)
+	c.mutex.Unlock()
+	return res
+}
+
+func (c *codes) Flush() error {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if len(c.pending) == 0 {
+		return nil
+	}
+	codes := make(map[common.Hash][]byte, len(c.pending))
+	for _, hash := range c.pending {
+		codes[hash] = c.codes[hash]
+	}
+
+	size, err := appendCodes(codes, c.file)
+	if err == nil {
+		c.pending = c.pending[:0]
+	}
+	c.fileSize = size
+	return err
+}
+
+func (c *codes) GetMemoryFootprint() *common.MemoryFootprint {
+	var sizeCodes uint
+	c.mutex.Lock()
+	for k, v := range c.codes {
+		sizeCodes += uint(len(k) + len(v))
+	}
+	c.mutex.Unlock()
+	return common.NewMemoryFootprint(unsafe.Sizeof(*c) + uintptr(sizeCodes))
```

Comment:
I wonder, why `unsafe.Sizeof(*c)` does not include the map, why do you need to count the `sizeCodes` 'by hand'?